geps/gep-3155/index.md

```diff
@@ -0,0 +1,205 @@
+
+## Introduction
+
+This is a wide ranging GEP intending to cover three additions to the API that all
+have a shared goal - enabling backend mTLS with Gateway API. Although this
+specific GEP focuses on manual configuration across the board, the hope is that
+it will also enable higher level automation to simplify this process for users.
+
+## API
+
+### Client Certs on Gateways
+
+A key requirement of mTLS is that the Gateway can provide a client cert to the
+backend. This adds that configuration to both Gateway and Service (via 
+BackendTLSPolicy).
+
+Specifying credentials at the service level allows to provide different
+credentials to connect to specific backend. This allows to address a use case,
+when destination service owner requires their clients to use a dedicated client
+certificate.
+
+**1. Add a new `BackendTLS` field at the top level of Gateways**
+
+```go
+type GatewaySpec struct {
+  // BackendTLS configures TLS settings for when this Gateway is connecting to
+  // backends with TLS.
+  BackendTLS GatewayBackendTLS `json:"backendTLS,omitempty"'
+}
+
+type GatewayBackendTLS struct {
+  // ClientCertificateRef is a reference to an object that contains a Client
+  // Certificate.
+  //
+  // References to a resource in different namespace are invalid UNLESS there
+  // is a ReferenceGrant in the target namespace that allows the certificate
+  // to be attached. If a ReferenceGrant does not allow this reference, the
+  // "ResolvedRefs" condition MUST be set to False for this listener with the
+  // "RefNotPermitted" reason.
+  //
+  // ClientCertificateRef can reference to standard Kubernetes resources, i.e.
+  // Secret, or implementation-specific custom resources.
+  //
+  // This setting can be overriden on the service level by use of BackendTLSPolicy.
+  ClientCertificateRef SecretObjectReference `json:"clientCertificateRef,omitempty"`
+}
+```
+
+**2. Add a new `Client` field at the top level of BackendTLSPolicy**
+
+```go
+type BackendTLSPolicySpec {
+  // Client specifies overriden TLS settings for the given service.
+  //
+  // If Client is omitted, backend mTLS settings inherited from the given gateway are
+  // used.
+  Client TLSClientSettings `json:"client"`
+}
+
+type TLSClientSettings struct {
+  // CertificateRef is a reference to an object that contains a Client
+  // Certificate.
+  //
+  // References to a resource in different namespace are invalid UNLESS there
+  // is a ReferenceGrant in the target namespace that allows the certificate
+  // to be attached. If a ReferenceGrant does not allow this reference, the
+  // "ResolvedRefs" condition MUST be set to False for this listener with the
+  // "RefNotPermitted" reason.
+  //
+  // CertificateRef can reference to standard Kubernetes resources, i.e.
+  // Secret, or implementation-specific custom resources.
+  //
+  // If CertificateRef is not specified, no client certificate is presented when
+  // connecting to the service.
```

Comment:
This doesn't quite seem accurate, as it seems a client certificate could be presented still if GatewayBackendTLS were configured?

I'm also not quite clear on the use case for allowing this to be specified in two different places, could you elaborate on that in the GEP please?

Reply:
The idea here is to allow overriding the default from GatewayBackendTLS with no certificate.

This is to support some compatibility corner cases, for example around a pre-packaged software.

Let's imagine a prepackaged software that always requests a certificate, but if customer refuses, it still proceeds with connection but falls back to an alternative authentication method, e.g. JWT. But for some reason if it observes an unexpected certificate it rejects the connection.

If we always presented the certificate in such case, it would be impossible to use this prepackaged software correctly, without creating a dedicated gateway with TLS disabled.

Having said that, if there is a consensus that this corner case is too far fetched, I am happy to drop it.

Reply:
Expecting an undefined or explicit `nil` here to override a configured value feels awkward and potentially confusing. It feels like the right way to handle this need would be to somehow scope GatewayBackendTLS better to only be applied for some backend connections?

Reply:
Agreed, replaced it with an Enum